jerry-core/ecma/operations/ecma-map-object.c

```diff
@@ -0,0 +1,757 @@
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+ecma_value_t
+ecma_op_map_has (ecma_value_t this_arg, /**< this argument */
+                 ecma_value_t key_arg, /**< key argument */
+                 bool is_weak_map) /**< map is weak map */
+{
+  ecma_map_object_t *map_object_p = ecma_op_map_get_object (this_arg, is_weak_map);
+  if (map_object_p == NULL)
+  {
+    return ECMA_VALUE_ERROR;
+  }
+
+  if (is_weak_map && !ecma_is_value_object (key_arg))
+  {
+    return ECMA_VALUE_FALSE;
+  }
+
+  return ecma_make_boolean_value (ecma_builtin_map_search (map_object_p->first_chunk_cp, key_arg) != NULL);
+} /* ecma_op_map_has */
+
+/**
+ * The generic map prototype object's 'set' routine
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+ecma_value_t
+ecma_op_map_set (ecma_value_t this_arg, /**< this argument */
+                 ecma_value_t key_arg, /**< key argument */
+                 ecma_value_t value_arg, /**< value argument */
+                 bool is_weak_map) /**< map is weak map */
+{
+  JERRY_UNUSED (key_arg);
+  JERRY_UNUSED (value_arg);
+
+  ecma_map_object_t *map_object_p = ecma_op_map_get_object (this_arg, is_weak_map);
+  if (map_object_p == NULL)
+  {
+    return ECMA_VALUE_ERROR;
+  }
+
+  if (is_weak_map && !ecma_is_value_object (key_arg))
+  {
+    return ecma_op_weak_map_key_error ();
+  }
+
+  ecma_value_t *value_p = ecma_builtin_map_search (map_object_p->first_chunk_cp, key_arg);
+
+  if (value_p == NULL)
+  {
+    ecma_value_t *key_p = NULL;
+    ecma_map_object_chunk_t *last_chunk_p = ECMA_GET_POINTER (ecma_map_object_chunk_t,
+                                                              map_object_p->last_chunk_cp);
+
+    if (last_chunk_p != NULL)
+    {
+      if (last_chunk_p->items[2] == ECMA_VALUE_ARRAY_HOLE)
+      {
+        key_p = last_chunk_p->items + 2;
+
+        if (last_chunk_p->items[1] == ECMA_VALUE_ARRAY_HOLE)
+        {
+          key_p = last_chunk_p->items + 1;
+          value_p = last_chunk_p->items + 2;
+        }
+      }
+    }
+
+    if (key_p == NULL || value_p == NULL)
+    {
+      size_t size = sizeof (ecma_map_object_chunk_t);
+      ecma_map_object_chunk_t *new_chunk_p = (ecma_map_object_chunk_t *) jmem_heap_alloc_block (size);
+
+      new_chunk_p->items[ECMA_MAP_OBJECT_ITEM_COUNT] = ecma_make_pointer_value(NULL);
+
+      for (int i = 0; i < ECMA_MAP_OBJECT_ITEM_COUNT; i++)
+      {
+        new_chunk_p->items[i] = ECMA_VALUE_ARRAY_HOLE;
+      }
+
+      ECMA_SET_NON_NULL_POINTER (map_object_p->last_chunk_cp, new_chunk_p);
+
+      if (last_chunk_p == NULL)
+      {
+        map_object_p->first_chunk_cp = map_object_p->last_chunk_cp;
+      }
+      else
+      {
+        last_chunk_p->items[ECMA_MAP_OBJECT_ITEM_COUNT] = ecma_make_pointer_value (new_chunk_p);
+      }
+
+      if (key_p == NULL)
+      {
+        JERRY_ASSERT (value_p == NULL);
+        key_p = new_chunk_p->items + 0;
```

Comment:
Does the `+0` is needed?

Reply:
Nope, but it does not hurt as well. I wanted the emphasize the zero value here, but it can be removed.